src/current/advisories/a104309.md

```diff
@@ -0,0 +1,87 @@
+---
+title: Technical Advisory a104309
+advisory: A-a104309
+summary: In rare cases, a rangefeed bug may cause a checkpoint to be emitted prematurely, before all writes below the checkpoint timestamp have been emitted.
+toc: true
+affected_versions: v2.1.11 to v22.2.17, v23.1.0 to v23.1.14, v23.2.0
+advisory_date: 2024-02-20
+docs_area: releases
+---
+
+Publication date: {{ page.advisory_date | date: "%B %e, %Y" }}
+
+## Description
+
+A [rangefeed](https://www.cockroachlabs.com/docs/stable/create-and-configure-changefeeds#enable-rangefeeds) bug may cause a [checkpoint](https://www.cockroachlabs.com/docs/stable/how-does-an-enterprise-changefeed-work) to be emitted prematurely in rare cases, before all writes below the checkpoint timestamp have been emitted. This could occur if nodes are overloaded or if long-running transactions (several seconds) are involved, such that intent resolution only replicates to a follower running a rangefeed more than 10 seconds after the transaction began. If a cluster using changefeeds experiences this bug, changefeeds will omit these write events, and the following error will be logged:
+
+{% include_cached copy-clipboard.html %}
+~~~ none
+cdc ux violation: detected timestamp ... that is less or equal to the local frontier
+~~~
+
+Across a large sample of clusters using changefeeds over a month, 3.8% of clusters had experienced this bug, and a total 0.000064% of events were omitted (one in a million).
+
+The bug is caused by a race condition that is typically triggered by transactions that run for several seconds or by follower replicas that lag significantly behind the leaseholder, such as when a node is overloaded or a network is disrupted. The rangefeed must be running on a follower replica that applies intent resolution long after the original transaction has committed and garbage-collected its intents and transaction record on the leaseholders (at least 10 seconds after the transaction begins).
+
+The specific conditions to trigger the bug are:
+
+1. An explicit or cross-range transaction commits, then asynchronously resolves all of its intents and removes its transaction record on all relevant leaseholders.
```

Comment:
Explicit OR cross-range transactions is a strange subset of all transactions. Is it explicit AND cross-range? 

Also, aren't all write intents removed async after the transaction commits during the cleanup stage? The language here just doesn't make sense to me.

Reply:
The way this is written reads (to me) as if this behavior is unusual.

Reply:
This might be worth clarifying. The intention here is "All explicit transactions and cross-range implicit transactions are affected."

Reply:
The reason is that single-range implicit transactions don't write intents.

Reply:
@shralex Is the suggested edit accurate? https://github.com/cockroachdb/docs/pull/18309/files#r1496690828 (sorry about the HTML, it's a tooling edge case)

Reply:
yes, LGTM!